dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala

```diff
@@ -172,10 +172,20 @@ class DdgGenerator(semantics: Semantics) {
       val identifierDestPairs =
         method._identifierViaContainsOut
           .flatMap { identifier =>
-            identifierToFirstUsages(identifier).map(usage => (identifier, usage))
+            val firstAndLastUsageByMethod = identifierToFirstUsages(identifier).groupBy(_.method)
+            firstAndLastUsageByMethod.values
+              .filter(_.nonEmpty)
+              .map(x => (x.head, x.last))
+              .flatMap { case (firstUsage, lastUsage) =>
+                (identifier.lineNumber, firstUsage.lineNumber, lastUsage.lineNumber) match {
+                  case (Some(iNo), Some(fNo), _) if iNo <= fNo => Some(identifier, firstUsage)
+                  case (Some(iNo), _, Some(lNo)) if iNo >= lNo => Some(lastUsage, identifier)
+                  case _                                       => None
+                }
+              }
           }
+          .toSet
```

Comment:
If the output is going to be a list in any case, best to use `distinct` instead of creating a set then a list.

Reply:
good catch

Reply:
@DavidBakerEffendi Have updated the PR